src/main/scala/akka/persistence/cassandra/query/journal/QueryActorPublisher.scala

```diff
@@ -0,0 +1,129 @@
+package akka.persistence.cassandra.query.journal
+
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ExecutionContext, Future}
+
+import akka.actor.{Actor, ActorLogging}
+import akka.pattern.pipe
+import akka.stream.actor.ActorPublisher
+import akka.stream.actor.ActorPublisherMessage.{Cancel, Request, SubscriptionTimeoutExceeded}
+
+//TODO: Optimizations - manage the buffer size efficienly, e.g. based on nr requests, remaining elements etc.
+//TODO: Database timeout, retry and failure handling.
+//TODO: Write tests for buffer size, delivery buffer etc.
+/**
+ * Abstract Query publisher. Can be integrated with Akka Streams as a Source.
+ * Intended to be extended by concrete Query publisher classes. This class manages the stream lifecycle,
+ * live stream updates, refreshInterval, max buffer size and causal consistency given an
+ * offset queryable data source. Causality is achieved by only a single request for data in flight
+ * at any point in time. Implementer must provide concrete initial state, update operation
+ * and end condition.
+ *
+ * @param refreshInterval Refresh interval.
+ * @param maxBufferSize Maximal buffer size.
+ * @tparam MessageType Type of message.
+ * @tparam State Type of state.
+ */
+private[journal] abstract class QueryActorPublisher[MessageType, State](
```

Comment:
I agree that we should extract common things to avoid duplication. I'm not so sure that using Futures is the best way, though. I think actors are better for this.